File: GoLang/0032_longest-valid-parentheses.go
/*=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 * 32. Longest Valid Parentheses
 * Hard
 * 
 * Given a string containing just the characters '(' and ')', find the length of the longest 
 * valid (well-formed) parentheses substring.
 * 
 * Example 1:
 * Input: "(()"
 * Output: 2
 * Explanation: The longest valid parentheses substring is "()"
 * 
 * Example 2:
 * Input: ")()())"
 * Output: 4
 * Explanation: The longest valid parentheses substring is "()()"
 * 
 ** =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=*/
 package problem0032

 func longestValidParentheses(s string) int {
	 var left, max, temp int
	 record := make([]int, len(s))
 

	 for i, b := range s {
		 if b == '(' {
			 left++
		 } else if left > 0 {
			 left--
			 record[i] = 2
		 }
	 }
 
	 
	 for i := 0; i < len(record); i++ {
		 if record[i] == 2 {
			 j := i - 1
			 for record[j] != 0 {
				 j--
			 }
			 record[i], record[j] = 1, 1
		 }
	 }
 
	 
	 for _, r := range record {
		 if r == 0 {
			 temp = 0
			 continue
		 }
 
		 temp++
		 if temp > max {
			 max = temp
		 }
	 }
 
	 return max
 }
 

File: Python/0032_longest-valid-parentheses.py
'''=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 * 32. Longest Valid Parentheses
 * Hard
 * 
 * Given a string containing just the characters '(' and ')', find the length of the longest 
 * valid (well-formed) parentheses substring.
 * 
 * Example 1:
 * Input: "(()"
 * Output: 2
 * Explanation: The longest valid parentheses substring is "()"
 * 
 * Example 2:
 * Input: ")()())"
 * Output: 4
 * Explanation: The longest valid parentheses substring is "()()"
 * 
 ** =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-='''
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        """
        :type s: str
        :rtype: int
        """
        stack, result = [-1], 0

        for index in range(len(s)):
        	if s[index] == '(':
        		stack.append(index)
        	else:
        		currIndex = stack.pop()
        		if currIndex == -1:
        			stack.append(index)
        		else:
        			result = max(result, index-currIndex+1)
        return result